network/notifiees.go

```diff
@@ -0,0 +1,77 @@
+package network
+
+import (
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// NotifyBundle implements Notifiee by calling any of the functions set on it,
+// and nop'ing if they are unset. This is the easy way to register for
+// notifications.
+type NotifyBundle struct {
+	ListenF      func(Network, ma.Multiaddr)
+	ListenCloseF func(Network, ma.Multiaddr)
+
+	ConnectedF    func(Network, Conn)
+	DisconnectedF func(Network, Conn)
+
+	OpenedStreamF func(Network, Stream)
+	ClosedStreamF func(Network, Stream)
+}
+
+var _ Notifiee = (*NotifyBundle)(nil)
+
+// Listen calls ListenF if it is not null.
+func (nb *NotifyBundle) Listen(n Network, a ma.Multiaddr) {
+	if nb.ListenF != nil {
+		nb.ListenF(n, a)
+	}
+}
+
+// ListenClose calls ListenCloseF if it is not null.
+func (nb *NotifyBundle) ListenClose(n Network, a ma.Multiaddr) {
+	if nb.ListenCloseF != nil {
+		nb.ListenCloseF(n, a)
+	}
+}
+
+// Connected calls ConnectedF if it is not null.
+func (nb *NotifyBundle) Connected(n Network, c Conn) {
+	if nb.ConnectedF != nil {
+		nb.ConnectedF(n, c)
+	}
+}
+
+// Disconnected calls DisconnectedF if it is not null.
+func (nb *NotifyBundle) Disconnected(n Network, c Conn) {
+	if nb.DisconnectedF != nil {
+		nb.DisconnectedF(n, c)
+	}
+}
+
+// OpenedStream calls OpenedStreamF if it is not null.
+func (nb *NotifyBundle) OpenedStream(n Network, s Stream) {
+	if nb.OpenedStreamF != nil {
+		nb.OpenedStreamF(n, s)
+	}
+}
+
+// ClosedStream calls ClosedStreamF if it is not null.
+func (nb *NotifyBundle) ClosedStream(n Network, s Stream) {
+	if nb.ClosedStreamF != nil {
+		nb.ClosedStreamF(n, s)
+	}
+}
+
+// Global noop notifiee. Do not change.
+var GlobalNoopNotifiee = &NoopNotifiee{}
+
+type NoopNotifiee struct{}
```

Comment:
I thought of that, but there is a nuisance, and it's totally subjective.

IMO `Noop` makes a strong concrete assertion that this will literally do nothing when called. `Null` is less strict: it conveys that _functionally speaking_ this is equivalent to having no connection manager. Underneath it may not noop, or it may actually do something, but its behaviour as externally observed is akin to a null object.

Granted, this is one man's opinion. And the benefit in DX of unifying naming may trump pragmatism. No strong opinions, but if we are to unify, I'd pick Noop.

Reply:
Yeah, I am fine with Noop, just want to take this chance to unify on something.